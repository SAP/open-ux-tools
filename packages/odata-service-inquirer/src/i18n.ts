import type { TOptions } from 'i18next';
import i18next from 'i18next';
import translations from './translations/odata-service-inquirer.i18n.json';

const odataServiceInquirerNamespace = 'odata-service-inquirer';
export const defaultProjectNumber = 1;
/**
 * Initialize i18next with the translations for this module.
 */
export async function initI18nOdataServiceInquirer(): Promise<void> {
    await i18next.init(
        {
            lng: 'en',
            fallbackLng: 'en',
            missingInterpolationHandler: () => ''
        },
        () => i18next.addResourceBundle('en', odataServiceInquirerNamespace, translations)
    );
}

/**
 * Helper function facading the call to i18next. Unless a namespace option is provided the local namespace will be used.
 *
 * @param key i18n key
 * @param options additional options
 * @returns {string} localized string stored for the given key
 */
export function t(key: string, options?: TOptions): string {
    if (!options?.ns) {
        options = Object.assign(options ?? {}, { ns: odataServiceInquirerNamespace });
    }
    return i18next.t(key, options);
}

initI18nOdataServiceInquirer().catch(() => {
    // Needed for lint
});
