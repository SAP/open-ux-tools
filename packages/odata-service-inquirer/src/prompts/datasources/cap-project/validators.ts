import { t } from '../../../i18n';
import { getCapProjectType } from '@sap-ux/project-access';

/**
 * Ensure the path specified is a valid CAP project.
 *
 * @param capProjectPath - The path to the CAP project
 * @returns A boolean indicating if the path is a valid CAP project or an error message
 */
export async function validateCapPath(capProjectPath: string): Promise<boolean | string> {
    if (capProjectPath) {
        try {
            return !!(await getCapProjectType(capProjectPath)) || t('prompts.validationMessages.capProjectNotFound');
        } catch (err) {
            return t('prompts.validationMessages.capProjectNotFound');
        }
    }
    return false;
}
