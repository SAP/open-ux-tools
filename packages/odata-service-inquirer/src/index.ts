import { type InquirerAdapter } from '@sap-ux/inquirer-common';
import { ToolsLogger, type Logger } from '@sap-ux/logger';
import { OdataVersion } from '@sap-ux/odata-service-writer';
import { ErrorHandler } from './error-handler/error-handler';
import { getQuestions } from './prompts';
import LoggerHelper from './prompts/logger-helper';
import { PromptStateHelper } from './prompts/prompt-helpers';
import { validateODataVersion } from './prompts/validators';
import type { OdataServiceAnswers, OdataServicePromptOptions, OdataServiceQuestion } from './types';
import { DatasourceType, promptNames, type CapRuntime, type CapService } from './types';
import { type ToolsSuiteTelemetryClient } from '@sap-ux/telemetry';
import { setTelemetryClient } from './utils';

/**
 * Get the inquirer prompts for odata service.
 *
 * @param promptOptions
 * @param logger    - a logger compatible with the {@link Logger} interface
 * @param enableGuidedAnswers
 * @returns the prompts used to provide input for odata service generation
 */
async function getPrompts(
    promptOptions?: OdataServicePromptOptions,
    logger?: Logger,
    enableGuidedAnswers = false
): Promise<OdataServiceQuestion[]> {
    // Initialize the logger refs
    LoggerHelper.logger = logger ?? new ToolsLogger({ logPrefix: '@sap-ux/odata-service-inquirer' });
    ErrorHandler.logger = LoggerHelper.logger;
    ErrorHandler.guidedAnswersEnabled = enableGuidedAnswers;

    return getQuestions(promptOptions);
}

/**
 * Prompt for odata service writer inputs.
 *
 * @param adapter
 * @param promptOptions
 * @param logger
 * @param enableGuidedAnswers - if true, the prompts will use guided answers to help users with validation errors
 * @param telemetryClient - the telemetry client to use for sending telemetry data
 * @returns the prompt answers
 */
async function prompt(
    adapter: InquirerAdapter,
    promptOptions?: OdataServicePromptOptions,
    logger?: Logger,
    enableGuidedAnswers?: boolean,
    telemetryClient?: ToolsSuiteTelemetryClient
): Promise<OdataServiceAnswers> {
    setTelemetryClient(telemetryClient);

    const odataServicePrompts = await getPrompts(promptOptions, logger, enableGuidedAnswers);

    const answers = await adapter.prompt<OdataServiceAnswers>(odataServicePrompts);

    // Add dervied service answers to the answers object
    Object.assign(answers, PromptStateHelper.odataService);

    return answers;
}

export {
    DatasourceType,
    OdataVersion,
    getPrompts,
    prompt,
    promptNames,
    // temp exports, remove once development is done
    validateODataVersion,
    type CapRuntime,
    type CapService,
    type InquirerAdapter,
    type OdataServiceAnswers,
    type OdataServicePromptOptions
};
