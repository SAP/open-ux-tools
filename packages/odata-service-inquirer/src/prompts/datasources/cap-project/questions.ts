import type { FileBrowserQuestion, ListQuestion, YUIQuestion } from '@sap-ux/inquirer-common';
import { OdataVersion } from '@sap-ux/odata-service-writer';
import { getCapCustomPaths } from '@sap-ux/project-access';
import type { Question } from 'inquirer';
import { t } from '../../../i18n';
import type {
    CapProjectChoice,
    CapProjectPaths,
    CapProjectRootPath,
    CapServiceAnswers,
    CapServiceChoice,
    OdataServicePromptAnswers,
    OdataServicePromptOptions
} from '../../../types';
import { PLATFORMS, internalPromptNames, promptNames } from '../../../types';
import { getPlatform } from '../../../utils';
import { PromptStateHelper, errorHandler } from '../../prompt-helpers';
import { enterCapPathChoiceValue, getCapEdmx, getCapProjectChoices, getCapServiceChoices } from './cap-helpers';
import { validateCapPath } from './validators';

/**
 * Find the specified choice in the list of CAP project choices and return its index.
 *
 * @param capChoices The list of CAP project choices
 * @param defaultChoicePath The path of the default choice
 * @returns The index of the default choice in the list of CAP project choices
 */
function getDefaultCapChoice(
    capChoices: CapProjectChoice[],
    defaultChoicePath?: string
): CapProjectChoice['value'] | number {
    if (defaultChoicePath) {
        return capChoices.findIndex((choice) => {
            if (typeof choice.value === 'string') {
                return choice.value === defaultChoicePath;
            }
            return choice.value.path === defaultChoicePath;
        });
    } else if (capChoices.length === 2) {
        return 0;
    }
    return -1;
}
/**
 * Get the prompts for selecting a CAP project from local path discovery.
 * Two prompts are returned, one for selecting a CAP project from a list of discovered projects and
 * one for entering a custom path to a CAP project.
 *
 * @param promptOptions - The prompt options which control CAP project search paths and default value
 * @returns the prompt used to provide input for selecting a CAP project
 */
export function getLocalCapProjectPrompts(
    promptOptions?: OdataServicePromptOptions
): (YUIQuestion<OdataServicePromptAnswers> | Question)[] {
    let capChoices: Awaited<CapProjectChoice[]> = [];
    const defaultCapPath = promptOptions?.[promptNames.capProject]?.defaultChoice;
    const defaultCapService = promptOptions?.[promptNames.capService]?.defaultChoice;
    let selectedCapProject: CapProjectPaths | undefined;
    let capServiceChoices: CapServiceChoice[];
    let defaultServiceIndex = 0;
    PromptStateHelper.reset();

    const prompts: (ListQuestion<CapServiceAnswers> | FileBrowserQuestion<CapServiceAnswers> | Question)[] = [
        {
            when: async (): Promise<boolean> => {
                capChoices = await getCapProjectChoices(promptOptions?.[promptNames.capProject]?.capSearchPaths ?? []);
                return capChoices?.length > 1;
            },
            type: 'list',
            name: promptNames.capProject,
            message: t('prompts.capProject.message'),
            default: () => {
                const defChoice = getDefaultCapChoice(capChoices, defaultCapPath);
                return defChoice;
            },
            choices: () => capChoices,
            guiOptions: {
                applyDefaultWhenDirty: true,
                mandatory: true,
                breadcrumb: t('prompts.capProject.breadcrumb')
            }
        } as ListQuestion<CapServiceAnswers>,
        {
            when: (answers): boolean => capChoices.length === 1 || answers?.capProject === enterCapPathChoiceValue,
            type: 'input',
            guiType: 'folder-browser',
            name: internalPromptNames.capProjectPath,
            message: t('prompts.capProjectPath.message'),
            default: () => {
                if (defaultCapPath) {
                    return defaultCapPath;
                }
            },
            guiOptions: { mandatory: true, breadcrumb: t('prompts.capProject.breadcrumb') },
            validate: async (projectPath: string): Promise<string | boolean> => {
                const validCapPath = await validateCapPath(projectPath);
                // Load the cap paths if the path is valid
                if (validCapPath === true) {
                    selectedCapProject = Object.assign(
                        { path: projectPath } as CapProjectRootPath,
                        await getCapCustomPaths(projectPath)
                    );
                    return true;
                }
                selectedCapProject = undefined;
                return validCapPath;
            }
        } as FileBrowserQuestion<CapServiceAnswers>,
        {
            when: async (answers) => {
                if (typeof answers?.capProject === 'object') {
                    selectedCapProject = answers.capProject;
                }

                if (selectedCapProject) {
                    capServiceChoices = await getCapServiceChoices(selectedCapProject);
                    return true;
                }
                return false;
            },
            type: 'list',
            name: promptNames.capService,
            message: t('prompts.capService.message'),
            choices: () => {
                if (defaultCapService) {
                    // Find the cap service, qualified by the provided project path
                    defaultServiceIndex = capServiceChoices?.findIndex(
                        (choice) =>
                            // project path is an extra validation that the prompt options for each
                            // of `defaultCapProject` and `defaultCapService` are compatible
                            choice.value?.projectPath === defaultCapService?.projectPath &&
                            choice.value?.serviceName === defaultCapService?.serviceName
                    );
                }
                return capServiceChoices;
            },
            guiOptions: {
                applyDefaultWhenDirty: true,
                mandatory: true,
                breadcrumb: true
            },
            default: () => {
                return capServiceChoices?.length > 1 ? defaultServiceIndex : 0;
            },
            validate: async (capService: CapServiceChoice['value']): Promise<string | boolean> => {
                const errMsg = errorHandler.getErrorMsg(undefined, true);
                if (errMsg) {
                    return errMsg;
                }
                if (capService) {
                    PromptStateHelper.odataService.metadata = await getCapEdmx(capService);
                    PromptStateHelper.odataService.servicePath = capService.urlPath;
                    PromptStateHelper.odataService.odataVersion = OdataVersion.v4;
                    return PromptStateHelper.odataService.metadata !== undefined
                        ? true
                        : t('prompts.validationMessages.metadataInvalid');
                }
                return false;
            }
        } as ListQuestion<CapServiceAnswers>
    ];

    if (getPlatform() === PLATFORMS.CLI) {
        prompts.push({
            when: async (answers: CapServiceAnswers): Promise<boolean> => {
                if (answers?.capService) {
                    PromptStateHelper.odataService.metadata = await getCapEdmx(answers?.capService);
                    PromptStateHelper.odataService.servicePath = answers?.capService.urlPath;
                    PromptStateHelper.odataService.odataVersion = OdataVersion.v4;
                }
                return false;
            },
            name: internalPromptNames.capCliStateSetter
        } as Question);
    }

    return prompts;
}
