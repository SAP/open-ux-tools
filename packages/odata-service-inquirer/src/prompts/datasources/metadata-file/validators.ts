/**
 * Validates the metadata file, returning an error string, if not valid, or an object with the valid metadata and version.
 *
 * @param path
 * @param odataVersion
 * @returns
 */

import { OdataVersion } from '@sap-ux/odata-service-writer';
import { readFile } from 'fs/promises';
import { validateODataVersion } from '../../validators';
import { t } from '../../../i18n';

export const validateMetadataFile = async (
    path: string,
    odataVersion?: OdataVersion
): Promise<
    | {
          version: OdataVersion;
          metadata: string;
      }
    | string
    | boolean
> => {
    if (!path) {
        return false;
    }

    try {
        let metadata = await readFile(path, 'utf-8');
        metadata = metadata.replace(/ & /g, ' &amp; ');
        const { validationMsg, version } = validateODataVersion(metadata, odataVersion);
        if (validationMsg) {
            return validationMsg;
        }
        return {
            version: version ?? OdataVersion.v4,
            metadata
        };
    } catch (error) {
        return t('prompts.validationMessages.metadataFilePathNotValid');
    }
};
