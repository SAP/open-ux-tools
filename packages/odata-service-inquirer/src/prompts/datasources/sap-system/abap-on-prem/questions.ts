import { withCondition } from '@sap-ux/inquirer-common';
import type { OdataVersion } from '@sap-ux/odata-service-writer';
import { validateClient } from '@sap-ux/project-input-validator';
import type { Answers, InputQuestion, PasswordQuestion, Question } from 'inquirer';
import { t } from '../../../../i18n';
import type { promptNames } from '../../../../types';
import {
    type OdataServiceAnswers,
    type OdataServicePromptOptions,
    type ServiceSelectionPromptOptions,
    type SystemNamePromptOptions
} from '../../../../types';
import { PromptState } from '../../../../utils';
import { ConnectionValidator } from '../../../connectionValidator';
import { getSystemServiceQuestion, getSystemUrlQuestion, getUserSystemNameQuestion } from '../new-system/questions';
import { newSystemPromptNames, type ServiceAnswer } from '../new-system/types';

const abapOnPremPromptNamespace = 'abapOnPrem';
const systemUrlPromptName = `${abapOnPremPromptNamespace}:${newSystemPromptNames.newSystemUrl}` as const;

export enum abapOnPremInternalPromptNames {
    sapClient = 'sapClient',
    systemUsername = 'abapSystemUsername',
    systemPassword = 'abapSystemPassword'
}

interface AbapOnPremAnswers extends Partial<OdataServiceAnswers> {
    [systemUrlPromptName]?: string;
    [abapOnPremInternalPromptNames.systemUsername]?: string;
    [abapOnPremInternalPromptNames.systemPassword]?: string;
    [promptNames.serviceSelection]?: ServiceAnswer;
}

/**
 * Get the Abap on-premise datasource questions.
 *
 * @param promptOptions options for prompts. Applicable options are: {@link ServiceSelectionPromptOptions}, {@link SystemNamePromptOptions}
 * @returns property questions for the Abap on-premise datasource
 */
export function getAbapOnPremQuestions(promptOptions?: OdataServicePromptOptions): Question<AbapOnPremAnswers>[] {
    PromptState.reset();
    const connectValidator = new ConnectionValidator();
    // Prompt options
    const requiredOdataVersion = promptOptions?.serviceSelection?.requiredOdataVersion;

    const questions: Question<AbapOnPremAnswers>[] = getAbapOnPremSystemQuestions(
        promptOptions?.userSystemName,
        connectValidator,
        requiredOdataVersion
    );

    questions.push(...getSystemServiceQuestion(connectValidator, abapOnPremPromptNamespace, promptOptions));

    return questions;
}

/**
 * Gets the Abap on-premise system questions.
 *
 * @param systemNamePromptOptions options for the system name prompt see {@link SystemNamePromptOptions}
 * @param connectionValidator reference to the existing connection validator, a new one will be created otherwise
 * @param requiredOdataVersion the required OData version for the service, this will be used to narrow the catalog service connections
 * @returns the Abap on-premise system questions
 */
export function getAbapOnPremSystemQuestions(
    systemNamePromptOptions?: SystemNamePromptOptions,
    connectionValidator?: ConnectionValidator,
    requiredOdataVersion?: OdataVersion
): Question<AbapOnPremAnswers>[] {
    const connectValidator = connectionValidator ?? new ConnectionValidator();
    let validClient = true;

    const questions: Question<AbapOnPremAnswers>[] = [
        getSystemUrlQuestion<AbapOnPremAnswers>(connectValidator, abapOnPremPromptNamespace, requiredOdataVersion),
        {
            type: 'input',
            name: abapOnPremInternalPromptNames.sapClient,
            message: t('prompts.sapClient.message'),
            guiOptions: {
                breadcrumb: t('prompts.sapClient.breadcrumb')
            },
            validate: (client) => {
                const valRes = validateClient(client);
                if (valRes === true) {
                    return (validClient = true);
                }
                validClient = false;
                return valRes;
            }
        } as InputQuestion<AbapOnPremAnswers>,
        {
            when: () => connectValidator.isAuthRequired(),
            type: 'input',
            name: abapOnPremInternalPromptNames.systemUsername,
            message: t('prompts.systemUsername.message'),
            guiOptions: {
                mandatory: true
            },
            default: '',
            validate: (user: string) => user?.length > 0
        } as InputQuestion<AbapOnPremAnswers>,
        {
            when: () => connectValidator.isAuthRequired(),
            type: 'password',
            guiOptions: {
                mandatory: true
            },
            name: abapOnPremInternalPromptNames.systemPassword,
            message: t('prompts.systemPassword.message'),
            guiType: 'login',
            mask: '*',
            default: '',
            validate: async (password, answers: AbapOnPremAnswers & Answers) => {
                if (!(connectValidator.validatedUrl && answers.abapSystemUsername && password && validClient)) {
                    return false;
                }
                const valResult = await connectValidator.validateAuth(
                    connectValidator.validatedUrl,
                    answers.abapSystemUsername,
                    password,
                    {
                        sapClient: answers.sapClient,
                        isSystem: true
                    }
                );
                if (valResult === true && connectValidator.serviceProvider) {
                    PromptState.odataService.connectedSystem = {
                        serviceProvider: connectValidator.serviceProvider
                    };
                    return true;
                }
                return valResult;
            }
        } as PasswordQuestion<AbapOnPremAnswers>
    ];

    if (systemNamePromptOptions?.hide !== true) {
        // New system question will allow user to give the system a user friendly name
        questions.push(
            withCondition(
                [getUserSystemNameQuestion(connectValidator, abapOnPremPromptNamespace)],
                (answers: AbapOnPremAnswers) =>
                    !!answers?.[systemUrlPromptName] &&
                    connectValidator.validity.reachable === true &&
                    (connectValidator.validity.authenticated || connectValidator.validity.authRequired !== true)
            )[0]
        );
    }

    return questions;
}
