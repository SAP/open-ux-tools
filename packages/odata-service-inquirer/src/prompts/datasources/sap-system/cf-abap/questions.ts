import type { Question } from 'inquirer';
import type { ServiceSelectionPromptOptions } from '../../../../types';
import { ConnectionValidator } from '../../../connectionValidator';
import { getCFDiscoverPrompts } from '../abap-on-btp/questions';
import { getSystemServiceQuestion } from '../service-selection/questions';
import type { ServiceAnswer } from '../service-selection/types';

/**
 * Cloud Foundry ABAP system prompts specifically for BAS environment since it requires additional destination configuration.
 * These will call out to cf tools to discover the available abap systems and services, and create a new destination if necessary,
 * to allow apps to access these cf hosted services.
 *
 * @param promptOptions prompt options to control some prompt behavior see {@link ServiceSelectionPromptOptions}
 * @returns the prompt questions
 */
export function getCfAbapBASQuestions(promptOptions?: ServiceSelectionPromptOptions): Question<ServiceAnswer>[] {
    // Using a prompt namespace allows re-use of system service (catalog based) selection prompt
    const cfAbapBasPromptNamespace = 'cfAbapBas';
    const connectionValidator = new ConnectionValidator();
    return [
        ...getCFDiscoverPrompts(connectionValidator, cfAbapBasPromptNamespace, promptOptions?.requiredOdataVersion),
        ...getSystemServiceQuestion(connectionValidator, cfAbapBasPromptNamespace, promptOptions)
    ];
}
