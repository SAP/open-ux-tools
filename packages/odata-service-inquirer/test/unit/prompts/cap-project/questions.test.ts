import type { ListQuestion } from 'inquirer';
import { initI18nOdataServiceInquirer, t } from '../../../../src/i18n';
import { getLocalCapProjectPrompts } from '../../../../src/prompts/datasources/cap-project/questions';
import type { CapService } from '../../../../src/types';
import {
    type CapProjectChoice,
    type CapServiceChoice,
    PLATFORMS,
    promptNames,
    internalPromptNames
} from '../../../../src/types';
import { getPlatform } from '../../../../src/utils';
import {
    enterCapPathChoiceValue,
    getCapServiceChoices as getCapServiceChoicesMock
} from '../../../../src/prompts/datasources/cap-project/cap-helpers';
import * as capValidators from '../../../../src/prompts/datasources/cap-project/validators';
import type { CapCustomPaths } from '@sap-ux/project-access';
import { PromptStateHelper, errorHandler } from '../../../../src/prompts/prompt-helpers';

jest.mock('../../../../src/utils', () => ({
    ...jest.requireActual('../../../../src/utils'),
    getPlatform: jest.fn()
}));

jest.mock('../../../../src/error-handler/error-handler', () => ({
    ...jest.requireActual('../../../../src/error-handler/error-handler'),
    getErrorMsg: jest.fn().mockImplementation(() => {
        return 'Any error message';
    })
}));

const mockCapCustomPaths: CapCustomPaths = {
    app: 'app',
    srv: 'src',
    db: 'db'
};

jest.mock('@sap-ux/project-access', () => ({
    ...jest.requireActual('../../../../src/error-handler/error-handler'),
    getCapCustomPaths: jest.fn().mockImplementation(async () => mockCapCustomPaths)
}));

const initialMockCapServicesChoices: CapServiceChoice[] = [
    {
        name: 'AdminService',
        value: {
            serviceName: 'AdminService',
            urlPath: '/admin/',
            serviceCdsPath: '../bookshop/srv/admin-service',
            projectPath: '/cap/path/to/project1'
        }
    },
    {
        name: 'CatalogService',
        value: {
            serviceName: 'CatalogService',
            urlPath: '/browse/',
            serviceCdsPath: '../bookshop/srv/cat-service',
            projectPath: '/cap/path/to/project1'
        }
    }
];

let mockCapServiceChoices: CapServiceChoice[] = initialMockCapServicesChoices;
let mockCapProjectChoices: CapProjectChoice[] = [];
let mockCapWorkspaceFolders: CapProjectChoice[] = [];

const initialMockEdmx = '<?xml version="1.0" encoding="utf-8"?><edmx:Edmx Version="2"/>';
let mockEdmx: string | undefined = initialMockEdmx;

jest.mock('../../../../src/prompts/datasources/cap-project/cap-helpers', () => ({
    ...jest.requireActual('../../../../src/prompts/datasources/cap-project/cap-helpers'),
    getCapServiceChoices: jest.fn().mockImplementation(async () => mockCapServiceChoices),
    getCapEdmx: jest.fn().mockImplementation(async () => mockEdmx),
    getCapProjectPaths: jest.fn().mockImplementation(() => mockCapWorkspaceFolders),
    getCapProjectChoices: jest.fn().mockImplementation(async () => mockCapProjectChoices)
}));

describe('getLocalCapProjectPrompts', () => {
    beforeAll(async () => {
        (getPlatform as jest.Mock).mockReturnValue(PLATFORMS.CLI);
        // Wait for i18n to bootstrap so we can test localised strings
        await initI18nOdataServiceInquirer();
    });

    beforeEach(() => {
        // Ensure each test is isolated, reset mocked function return values to initial state
        mockCapServiceChoices = initialMockCapServicesChoices;
        mockCapProjectChoices = [];
        mockCapWorkspaceFolders = [];
        mockEdmx = initialMockEdmx;
    });

    test('getLocalCapProjectPrompts, returns expected prompts', async () => {
        const capPrompts = await getLocalCapProjectPrompts();
        // Verify static values
        expect(capPrompts).toMatchSnapshot();
    });

    test('prompt: capProject', async () => {
        let capPrompts = await getLocalCapProjectPrompts();
        let capProjectPrompt = capPrompts.find((prompt) => prompt.name === promptNames.capProject);
        if (!capProjectPrompt) {
            fail('`capProjectPrompt` is not defined');
        }
        // Dont show the project selection prompt if only one entry ('Enter path manually') since we will prompt for the path by default
        expect(await (capProjectPrompt.when as Function)()).toEqual(false);

        mockCapProjectChoices = [
            {
                name: 'project1',
                value: {
                    folderName: 'project1',
                    path: '/abs/path/to/cap/project1',
                    app: 'app',
                    db: 'db',
                    srv: 'srv'
                }
            },
            {
                name: t('prompts.capProject.enterCapPathChoiceName'),
                value: 'enterCapPath'
            }
        ];
        // Show the project selection prompt when more than 1 choice
        expect(await (capProjectPrompt.when as Function)()).toEqual(true);
        expect(((capProjectPrompt as ListQuestion).choices as Function)().length).toEqual(2);
        expect(((capProjectPrompt as ListQuestion).choices as Function)()).toEqual(mockCapProjectChoices);

        // find the default cap project choice, 2 choices (one project), always pre-select (0) to avoid selection message
        expect((capProjectPrompt.default as Function)()).toEqual(0);
        mockCapProjectChoices.unshift({
            name: 'project2',
            value: {
                folderName: 'project2',
                path: '/abs/path/to/cap/project2',
                app: 'app',
                db: 'db',
                srv: 'srv'
            }
        });
        expect((capProjectPrompt.default as Function)()).toEqual(-1);

        capPrompts = await getLocalCapProjectPrompts({
            capProject: {
                capSearchPaths: [],
                defaultChoice: '/abs/path/to/cap/project1'
            }
        });
        capProjectPrompt = capPrompts.find((prompt) => prompt.name === promptNames.capProject);
        // It is expect that the `when` condition must first execute to preload the cap choices
        expect(await (capProjectPrompt!.when as Function)()).toEqual(true);
        expect(((capProjectPrompt as ListQuestion).choices as Function)().length).toEqual(3);
        expect(((capProjectPrompt as ListQuestion).choices as Function)()).toEqual(mockCapProjectChoices);
        expect((capProjectPrompt!.default as Function)()).toEqual(1);
    });

    test('prompt: capProjectPath', async () => {
        let capPrompts = await getLocalCapProjectPrompts();
        let capProjectPathPrompt = capPrompts.find((prompt) => prompt.name === internalPromptNames.capProjectPath);

        // No previous answers, no cap project choices (from prompt `capProject`)
        expect(await (capProjectPathPrompt!.when as Function)()).toEqual(false);
        // Previous answer to enter manual path
        expect(
            await (capProjectPathPrompt!.when as Function)({ [promptNames.capProject]: enterCapPathChoiceValue })
        ).toEqual(true);
        // No previous choices, so only option is to enter manual path
        mockCapProjectChoices = [
            {
                name: t('prompts.capProject.enterCapPathChoiceName'),
                value: 'enterCapPath'
            }
        ];
        const capProjectPrompt = capPrompts.find((prompt) => prompt.name === promptNames.capProject);
        await (capProjectPrompt!.when as Function)();
        expect(await (capProjectPathPrompt!.when as Function)()).toEqual(true);

        capPrompts = await getLocalCapProjectPrompts({
            capProject: {
                capSearchPaths: [],
                defaultChoice: '/abs/path/to/cap/project1'
            }
        });
        capProjectPathPrompt = capPrompts.find((prompt) => prompt.name === internalPromptNames.capProjectPath);
        expect(await (capProjectPathPrompt!.default as Function)()).toEqual('/abs/path/to/cap/project1');

        // Validate
        expect(await (capProjectPathPrompt!.validate as Function)()).toEqual(false);
        const validateCapPathSpy = jest.spyOn(capValidators, 'validateCapPath').mockResolvedValue(true);
        expect(await (capProjectPathPrompt!.validate as Function)('/any/valid/cap/path')).toEqual(true);
        expect(validateCapPathSpy).toHaveBeenCalledWith('/any/valid/cap/path');

        jest.spyOn(capValidators, 'validateCapPath').mockResolvedValue('not valid');
        expect(await (capProjectPathPrompt!.validate as Function)('/not/valid/cap/path')).toEqual('not valid');
    });

    test('prompt: capService', async () => {
        let capPrompts = await getLocalCapProjectPrompts();
        let capServicePrompt = capPrompts.find((prompt) => prompt.name === promptNames.capService);
        (getCapServiceChoicesMock as jest.Mock).mockClear();
        const capProject = {
            path: '/cap/path/to/project1',
            folderName: 'project1',
            app: 'app',
            srv: 'srv',
            db: 'db'
        } as CapProjectChoice['value'];

        expect(await (capServicePrompt!.when as Function)()).toEqual(false);
        // Previously set cap project (via prompt options in this case)
        expect(
            await (capServicePrompt!.when as Function)({
                capProject
            })
        ).toEqual(true);
        expect(getCapServiceChoicesMock as jest.Mock).toHaveBeenCalledWith(capProject);
        expect(((capServicePrompt as ListQuestion).choices as Function)()).toEqual(initialMockCapServicesChoices);
        expect((capServicePrompt!.default as Function)()).toEqual(0);

        // Default selection when default service provided should be the default service choice entry
        capPrompts = await getLocalCapProjectPrompts({
            capService: {
                defaultChoice: {
                    serviceName: 'CatalogService',
                    projectPath: '/cap/path/to/project1'
                }
            }
        });
        capServicePrompt = capPrompts.find((prompt) => prompt.name === promptNames.capService);
        await (capServicePrompt!.when as Function)({ capProject });
        ((capServicePrompt as ListQuestion).choices as Function)();
        expect((capServicePrompt!.default as Function)()).toEqual(1);

        // validate (only run in YUI)
        expect(await (capServicePrompt!.validate as Function)()).toBe(false);
        // previous errors (from earlier cap prompts) are reported
        errorHandler.logErrorMsgs('Something unexpected happened!');
        expect(await (capServicePrompt!.validate as Function)()).toBe('Something unexpected happened!');
        // PromptStateHelper has the correct values after successful Cap service selection
        const capService: CapService = {
            projectPath: '/some/cap/project',
            serviceName: 'capService1',
            appPath: 'app',
            capType: 'Node.js',
            serviceCdsPath: '../relative/services/path',
            urlPath: 'odatav4/service/path'
        };
        expect(await (capServicePrompt!.validate as Function)(capService)).toBe(true);
        expect(PromptStateHelper.odataService).toMatchInlineSnapshot(`
            {
              "metadata": "<?xml version="1.0" encoding="utf-8"?><edmx:Edmx Version="2"/>",
              "odataVersion": "4",
              "servicePath": "odatav4/service/path",
            }
        `);

        // Invalid Cap edmx returned by `getCapEdmx` is handled
        mockEdmx = undefined;
        expect(await (capServicePrompt!.validate as Function)(capService)).toEqual(
            t('prompts.validationMessages.metadataInvalid')
        );
    });

    test('hidden prompt to set state on CLI - since list validators dont run there', async () => {
        const capPrompts = await getLocalCapProjectPrompts();
        const capCliStateSetterPrompt = capPrompts.find(
            (prompt) => prompt.name === internalPromptNames.capCliStateSetter
        );
        expect(await (capCliStateSetterPrompt!.when as Function)()).toEqual(false);
        expect(PromptStateHelper.odataService).toEqual({});

        const capService: CapService = {
            projectPath: '/some/cap/project',
            serviceName: 'capService1',
            appPath: 'app',
            capType: 'Node.js',
            serviceCdsPath: '../relative/services/path',
            urlPath: 'odatav4/service/path'
        };
        expect(await (capCliStateSetterPrompt!.when as Function)({ capService })).toEqual(false);
        expect(PromptStateHelper.odataService).toMatchInlineSnapshot(`
            {
              "metadata": "<?xml version="1.0" encoding="utf-8"?><edmx:Edmx Version="2"/>",
              "odataVersion": "4",
              "servicePath": "odatav4/service/path",
            }
        `);
    });
});
