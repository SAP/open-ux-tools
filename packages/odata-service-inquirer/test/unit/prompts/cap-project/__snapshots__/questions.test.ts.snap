// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLocalCapProjectPrompts getLocalCapProjectPrompts, returns expected prompts 1`] = `
[
  {
    "choices": [Function],
    "default": [Function],
    "guiOptions": {
      "applyDefaultWhenDirty": true,
      "breadcrumb": "CAP Project",
      "mandatory": true,
    },
    "message": "Choose your CAP project",
    "name": "capProject",
    "type": "list",
    "when": [Function],
  },
  {
    "default": [Function],
    "guiOptions": {
      "breadcrumb": "CAP Project",
      "mandatory": true,
    },
    "guiType": "folder-browser",
    "message": "CAP project folder path",
    "name": "capProjectPath",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  {
    "choices": [Function],
    "default": [Function],
    "guiOptions": {
      "applyDefaultWhenDirty": true,
      "breadcrumb": true,
      "mandatory": true,
    },
    "message": "OData service",
    "name": "capService",
    "type": "list",
    "validate": [Function],
    "when": [Function],
  },
  {
    "name": "capCliStateSetter",
    "when": [Function],
  },
]
`;
